packages/compass-schema/src/stores/store.ts

```diff
@@ -0,0 +1,370 @@
+      // Use a function here to a) ensure that the calculations here
+      // are only made when telemetry is enabled and b) that errors from
+      // those calculations are caught and logged rather than displayed to
+      // users as errors from the core schema analysis logic.
+      const trackEvent = () => ({
+        with_filter: Object.entries(this.query.filter).length > 0,
+        schema_width: schema?.fields?.length ?? 0,
+        schema_depth: schema ? calculateSchemaDepth(schema) : 0,
+        geo_data: schema ? schemaContainsGeoData(schema) : false,
+        analysis_time_ms: analysisTimeMS,
+      });
+      track('Schema Analyzed', trackEvent);
+    },
+
+    startAnalysis: async function (this: SchemaStore) {
+      const query = this.query || {};
+
+      const sampleSize = query.limit
+        ? Math.min(DEFAULT_SAMPLE_SIZE, query.limit)
+        : DEFAULT_SAMPLE_SIZE;
+
+      const samplingOptions = {
+        query: query.filter,
+        size: sampleSize,
+        fields: query.project ?? undefined,
+      };
+
+      const driverOptions = {
+        maxTimeMS: capMaxTimeMSAtPreferenceLimit(query.maxTimeMS),
+      };
+
+      try {
+        debug('analysis started');
+
+        const abortController = new AbortController();
+        const abortSignal = abortController.signal;
+
+        this.setState({
+          analysisState: ANALYSIS_STATE_ANALYZING,
+          errorMessage: '',
+          outdated: false,
+          schema: null,
+          abortController,
+        });
+
+        const analysisStartTime = Date.now();
+        const schema = await analyzeSchema(
+          this.dataService,
+          abortSignal,
+          this.ns,
+          samplingOptions,
+          driverOptions,
+          loggerAndTelemetry
+        );
+        const analysisTime = Date.now() - analysisStartTime;
+
+        this.setState({
+          analysisState: schema
+            ? ANALYSIS_STATE_COMPLETE
+            : ANALYSIS_STATE_INITIAL,
+          schema: schema,
+          resultId: resultId(),
+        });
+
+        this._trackSchemaAnalyzed(analysisTime);
+
+        this.onSchemaSampled();
+      } catch (err: any) {
+        log.error(
+          mongoLogId(1_001_000_188),
+          'Schema analysis',
+          'Error sampling schema',
+          {
+            error: err.stack,
+          }
+        );
+        this.setState({ ...getErrorState(err), resultId: resultId() });
+      } finally {
+        this.setState({ abortController: undefined });
+      }
+    },
+
+    storeDidUpdate(this: SchemaStore, prevState: SchemaState) {
+      debug('schema store changed from', prevState, 'to', this.state);
+    },
+  }) as SchemaStore;
+
+  on(localAppRegistry, 'subtab-changed', (name) => {
+    store.onSubTabChanged(name);
+  });
+
+  /**
+   * When the collection is changed, update the store.
+   */
+  on(localAppRegistry, 'query-changed', (state) => {
+    store.onQueryChanged(state);
+  });
```

Comment:
Seems like the new activate behavior doesn't map exactly to the old one and causes issues here (I was looking into why the e2e test was failing): we don't activate until this subtab is rendered for the first time and so it doesn't start listening to query changed until then causing the state to be out of sync initially. I am planning to remove this indirection in the query bar interface update, but for now I guess we can just pick the value from the query bar store on the first render similar to what schema tab UI is doing 🤔

Reply:
Yeah, didn’t see your comment and found out the same thing :) Luckily that’s easy enough to work around